Introduce custom settings for password recommendations

flo-mic opened the pull request:
I would like to move from lastpass to nextcloud passwords for serval reasons. I find the implementation really nice so thanks to all contributors here!

For me there is one importand setting missing and it seams that I'm not alone. If you have an ldap server you will probably have multiple web pages with the same user. With the currend implementation of password recommendations this will get a problem. For this reason I implemented a possibility to change the way how password recommendations are searched.

In the advanced extension settings a new section "Password recommendations" was added. You can define the following settings there.

1. **Search passwords based on the following option**. This defines the way password recommendations are searched. The setting has 4 options (default is "host"). Below is some example to explain how they work when you are on the web page "www.example.com:8443/login":
 
   - **Domain** - This will search for all passwords on the same domain. 
       - example.com                             -> visible
       - www.example.com                    -> visible
       - www.example.com:8443            -> visible
       - www.example.com:8443/login  -> visible
     
   - **Host** - This will search for all passwords on the same host. 
       - example.com                             -> not visible
       - www.example.com                    -> visible
       - www.example.com:8443            -> visible
       - www.example.com:8443/login  -> visible
       
   - **Host + Port** - This will search for all passwords on the same host and port. 
       - example.com                             -> not visible
       - www.example.com                    -> not visible
       - www.example.com:8443            -> visible
       - www.example.com:8443/login  -> visible
       
   - **Exact url** - This will search for all passwords where the url is equal to the current web page. 
       - example.com                             -> not visible
       - www.example.com                    -> not visible
       - www.example.com:8443            -> not visible
       - www.example.com:8443/login  -> visible

2. **Maximum number of results for the password recommendation**. This defines how many results are shown. I found this #141 open issue. Due to the fact that I was customizing this class for the first setting it was an easy change to add also the second option. You can define a value between 1 and 20 (default is set to 8).

![image](https://user-images.githubusercontent.com/52607335/108595243-15e06380-737f-11eb-97a2-7dcd19144bee.png)

The following issues can be fixed with this change:
Fixes #141
Fixes #87
Fixes #14

This change also replaces #147 because it allows configuring this setting in the extension settings.

marius-wieschollek commented on src/js/Controller/Setting/Get.js:
```diff
@@ -20,7 +20,9 @@ export default class Get extends AbstractController {
             'theme.custom',
-            'debug.localisation.enabled'
+            'debug.localisation.enabled',
+            'search.recommendation.option',
```
It's hard to find a good name for this setting, but "option" just says nothing to me.
Maybe something like "search.recommendation.mode" or "search.recommendation.base" would be more fitting.

flo-mic replied:
I agree with you the naming was bad. I changed it to "search.recommendation.mode". But I'm not sure if maybe "search.recommendation.filter" would be even better. What do you think?

marius-wieschollek replied:
The reason i suggested "mode" is that we could probably add something like an automatic mode or some other kind of more advanced recommendation later and then this setting name would still make sense.

flo-mic replied:
ok sounds logical for me. Thanks for the explanation.

marius-wieschollek commented on src/js/Manager/RecommendationManager.js:
```diff
@@ -66,12 +84,22 @@ class RecommendationManager {
+                            "")
+                        )
+                    )
+                )
```
This is really hard to read. It would probably be better to pass the query to a different function here that returns the actual filter. That function can then have ifs for all the different options.

```
        query
            .where(...this._getFilterQuery(this.options.searchQuery, query))

[...]

_getFilterQuery(mode, query) {
  if(mode === 'host') {
    return [
      query.field('host').equals(url.host)
    ]
  }
  [...]
}
```

flo-mic replied:
I moved the filter expression in a seperate function as requested.

marius-wieschollek commented on src/js/Manager/RecommendationManager.js:
```diff
@@ -83,6 +111,20 @@ class RecommendationManager {
+     * @param {String} host
+     */
+    getSearchDomainFromHost(host) {
+        var regIPAdress = /((^\s*((([0-9]|[1-9][0-9]|1[0-9]{2}|2[0-4][0-9]|25[0-5])\.){3}([0-9]|[1-9][0-9]|1[0-9]{2}|2[0-4][0-9]|25[0-5]))\s*$)|(^\s*((([0-9A-Fa-f]{1,4}:){7}([0-9A-Fa-f]{1,4}|:))|(([0-9A-Fa-f]{1,4}:){6}(:[0-9A-Fa-f]{1,4}|((25[0-5]|2[0-4]\d|1\d\d|[1-9]?\d)(\.(25[0-5]|2[0-4]\d|1\d\d|[1-9]?\d)){3})|:))|(([0-9A-Fa-f]{1,4}:){5}(((:[0-9A-Fa-f]{1,4}){1,2})|:((25[0-5]|2[0-4]\d|1\d\d|[1-9] ...
```
I don't understand it, but it works! I'm amazed!

flo-mic replied:
This regex checks if the input host is an ipv4 or ipv6 address. In both cases it will return the host and will not try to extract the domain. This is importand because it makes no sense to extract domain from a ip address. And if you have an ipv6 address as url the code below would cause errors.

marius-wieschollek replied:
:+1:

marius-wieschollek commented on src/platform/generic/_locales/de/messages.json:
```diff
@@ -1176,5 +1176,117 @@
+    },    
+    "LabelSearchRecommendationRows20"                 : {
+        "message"    : "20",
+        "description": "Number of results that should be displayed for password recommendations."
```
We can't do that to our translators. See below for suggestion.

flo-mic replied:
I agree an dI was also not happy with my implementation. I tried befor with the id a slabel but this caused some errors in the console log. So I used this workaround. 

But your appreach is of course better. I updated my implementation according your instructions.

marius-wieschollek commented on src/vue/Components/Options/Settings.vue:
```diff
@@ -48,27 +59,61 @@
+                var i = 1;
+                var result = [];
+                for(i =1; i <= 20; i++) {
+                    result.push({id: i, label: "LabelSearchRecommendationRows" + i.toString()})
```
You can pass an array as label and then use a variable in the language tag.

See https://github.com/marius-wieschollek/passwords-webextension/blob/master/src/vue/Components/Accounts/Account.vue#L65 for a similar use case.

marius-wieschollek reviewed (changes requested):
Thanks for your work, recommendations really seem to be a bit of a touchy issue so far.

flo-mic commented:
All your recommendatiosn are applied. And I also added a new class which i missed in the initial pull request. The change now includes a "FieldStartsWith" class which I added to the "FieldFactory". This is required when listing all passwords that math the "host" condition without checking for port or pathname.

flo-mic commented:
I just added the recommendations settings to the password MiningManager. So if I use e.g. domain as search condition it will not recognize a password again if i use this already on another subdomain. Also if an update was detected it will use the recommendations search condition to detect the the right entry for the update.

@marius-wieschollek Please let me knwo if anything is missing or if this is an consistent approach that can be merged.

marius-wieschollek commented:
I made one more change before deploying it to the development build.
I switched from `SettingsService.getValue` to `SettingsService.get` which returns a setting model that always has the current value of the setting. That removes the need to restart the extension after changing the setting. The new value will then just be applied the next time the recommendations are fetched.
